src/YesSql.Core/Session.cs

```diff
@@ -474,6 +471,10 @@ public void Delete(object obj, string collection = null)
 
             var state = GetState(collection);
 
+            // First detach the object from the session incase it was previously added by other
+            // methods like SaveAsync or UpdateEntityAsync.
+            DetachInternal(obj, state);
```

Comment:
Maybe it should distinguish between a brand new item and and an updated one, because the with a brand new one it should not be added to the Delete collection. This would also ensure that adding it again after the Delete call would work (test to be added).

The test should also check that the indexes are not computed if the object is forgotten.